rpc/json-rpc/src/response.rs

```diff
@@ -0,0 +1,402 @@
+    /// let json = serde_json::to_string(&ok).unwrap();
+    /// assert_eq!(json, r#"{"jsonrpc":"2.0","id":0,"error":{"code":-32600,"message":"Invalid Request"}}"#);
+    /// ```
+    pub const fn invalid_request(id: Id) -> Self {
+        Self {
+            jsonrpc: Version,
+            payload: Err(ErrorObject::invalid_request()),
+            id,
+        }
+    }
+
+    /// Creates an error response using [`ErrorObject::method_not_found`].
+    ///
+    /// ```rust
+    /// use json_rpc::{Id, Response, error::{ErrorObject, ErrorCode}};
+    ///
+    /// let ok = Response::<()>::method_not_found(Id::Num(0));
+    /// let json = serde_json::to_string(&ok).unwrap();
+    /// assert_eq!(json, r#"{"jsonrpc":"2.0","id":0,"error":{"code":-32601,"message":"Method not found"}}"#);
+    /// ```
+    pub const fn method_not_found(id: Id) -> Self {
+        Self {
+            jsonrpc: Version,
+            payload: Err(ErrorObject::method_not_found()),
+            id,
+        }
+    }
+
+    /// Creates an error response using [`ErrorObject::invalid_params`].
+    ///
+    /// ```rust
+    /// use json_rpc::{Id, Response, error::{ErrorObject, ErrorCode}};
+    ///
+    /// let ok = Response::<()>::invalid_params(Id::Num(0));
+    /// let json = serde_json::to_string(&ok).unwrap();
+    /// assert_eq!(json, r#"{"jsonrpc":"2.0","id":0,"error":{"code":-32602,"message":"Invalid params"}}"#);
+    /// ```
+    pub const fn invalid_params(id: Id) -> Self {
+        Self {
+            jsonrpc: Version,
+            payload: Err(ErrorObject::invalid_params()),
+            id,
+        }
+    }
+
+    /// Creates an error response using [`ErrorObject::internal_error`].
+    ///
+    /// ```rust
+    /// use json_rpc::{Id, Response, error::{ErrorObject, ErrorCode}};
+    ///
+    /// let ok = Response::<()>::internal_error(Id::Num(0));
+    /// let json = serde_json::to_string(&ok).unwrap();
+    /// assert_eq!(json, r#"{"jsonrpc":"2.0","id":0,"error":{"code":-32603,"message":"Internal error"}}"#);
+    /// ```
+    pub const fn internal_error(id: Id) -> Self {
+        Self {
+            jsonrpc: Version,
+            payload: Err(ErrorObject::internal_error()),
+            id,
+        }
+    }
+}
+
+//---------------------------------------------------------------------------------------------------- Trait impl
+impl<T> std::fmt::Display for Response<T>
+where
+    T: Clone + Serialize,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match serde_json::to_string_pretty(self) {
+            Ok(json) => write!(f, "{json}"),
+            Err(_) => Err(std::fmt::Error),
+        }
+    }
+}
+
+//---------------------------------------------------------------------------------------------------- Key
+/// This type represents the `"key"` values within [`Response`].
+///
+/// We need this since `Response` has a custom deserializer implementation.
+pub(crate) enum Key {
```

Comment:
Why does this need a custom serde impl?

Reply:
Removed with addition of `enum ResponsePayload`.